sources/constraint-declaration-validation.asciidoc

```diff
@@ -1026,6 +1100,289 @@ public class SimpleOrderService extends OrderService {
+* the `@ValidAddress` constraint will be applied to all elements of all lists stored in the map
+
+Container element constraints must not be put to wildcard type arguments (with or without bounds), i.e. the following usage is disallowed:
+
+.Disallowed usage of a container element constraint on a wildcard type argument
+====
+[source, JAVA]
+----
+private List<@NotEmpty ?> list;
+----
+====
+
+[[constraintdeclarationvalidationprocess-containerelementconstraints-implicitunwrapping]]
+==== Implicit unwrapping of containers
+
+Besides specifying container element constraints on type arguments, it is also possible to declare container element constraints on non-generic container types.
+This is done by means of _implicit unwrapping_, i.e. a constraint doesn't apply to the annotated container itself but to its elements.
+An example are JavaFX's non-generic property types such as `StringProperty` or `IntegerProperty`:
+
+.Implicit unwrapping of container elements
+====
+[source, JAVA]
+----
+@Size(max=10) private StringProperty name;
+@Positive private IntegerProperty positiveInt;
+----
+====
+
+Here the `@Size` and `@Positive` constraints don't apply to the annotated `StringProperty` and `IntegerProperty` themselves, but rather to the wrapped string and integer values, respectively.
+
+For this to work, an unambiguously resolvable value extractor must be defined which carries the `@UnwrapByDefault` annotation (see <<constraintdeclarationvalidationprocess-valueextractordefinitions-unwrapbydefault>>).
+
+If needed, the target (container or container element) of a constraint declared on a container can be explicitly specified via the `Unwrap` and `Skip` payload definitions:
+
+.`Payload` types for unwrapping control
+====
+[source, JAVA, indent=0]
+----
+include::{validation-api-source-dir}javax/validation/valueextraction/Unwrapping.java[lines=7..8;11..-1]
+----
+====
+
+This is useful for applying a constraint given on a non-generic container to
+
+* the container element if there is no value extractor marked with `@UnwrapByDefault` (by using `Unwrap`)
+* the container itself in case there is a value extractor marked with `@UnwrapByDefault` (by using `Skip`)
+
+For instance the `@NotNull` constraint is applied to the `StringProperty` container:
+
+[source,java]
+----
+@NotNull(payload = Unwrapping.Skip.class) StringProperty name;
+----
+
+[NOTE]
+====
+For the sake of readability, when applying constraints to the elements of a generic container,
+it is strongly recommended to put the constraints to the type argument instead of the container itself in conjunction with `Unwrapping.Unwrap`.
+I.e. you should prefer
+
+[source,java]
+----
+List<@Email String> emails;
+----
+
+over
+
+[source,java]
+----
+@Email(payload = Unwrapping.Unwrap.class)
+List<String> emails;
+----
+====
+
+[[constraintdeclarationvalidationprocess-valueextractordefinitions]]
+=== Value extractor definitions
```

Comment:
I considered it, but in "Constraint Definition" it seemed wrong. We can do a separate chapter, although it seems a bit heavy.